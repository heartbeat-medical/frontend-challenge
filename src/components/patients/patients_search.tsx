import { FunctionComponent, useState, useEffect } from "react";
import { Patient, PatientSearchQuery } from "./patients";
import { Loading } from "../loading/loading";

type props = {
  loadPatients: (query: PatientSearchQuery) => Promise<Patient[]>;
  onResults: (ps: Patient[]) => void;
};

export const PatientsSearch: FunctionComponent<props> = ({
  loadPatients,
  onResults,
}) => {
  const [query, updateQuery] = useState("");
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    const makeRequest = () => {
      setIsLoading(true);
      const sq: PatientSearchQuery = {
        name: query,
        ehrID: query,
        id: query,
      };
      
      loadPatients(sq)
        .then((ps) => {
          onResults(ps);
        })
        .catch((err) => alert(err))
        .finally(() => {
          setIsLoading(false);
        })
    };

    makeRequest();
  }, [query, loadPatients, onResults])


  const handleChange = (e: any) => {
    updateQuery(e.target.value);
  }

  return (
    <div>
      {isLoading && <Loading />}
      <input
        onChange={handleChange}
      />
    </div>
  );
};

type psearchboxprops = {
  onQueryChange: (query: PatientSearchQuery) => void;
};
