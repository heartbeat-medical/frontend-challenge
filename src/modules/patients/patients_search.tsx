import { FunctionComponent as FC, useState } from "react";
import { Patient, PatientSearchQuery } from "./patients";
import { ToastBox, Loader } from "../../components";
import "./patients.css";

type props = {
  loadPatients: (query: PatientSearchQuery) => Promise<Patient[]>;
  onResults: (ps: Patient[]) => void;
};

const PatientsSearch: FC<props> = ({ loadPatients, onResults }) => {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState("");
  const makeRequest = (inputValue: string) => {
    setIsLoading(true);
    const sq: PatientSearchQuery = {
      name: inputValue,
      ehrID: inputValue,
      id: inputValue,
    };
    loadPatients(sq)
      .then((ps) => {
        setIsLoading(false);
        setError("");
        onResults(ps);
      })
      .catch((err) => {
        setIsLoading(false);
        setError(err);
      });
  };
  return (
    <>
      <form>
        <label htmlFor="search-user">
          <input
            name="search-user"
            placeholder="Search by name, emp ID"
            onChange={(e) => makeRequest(e.target.value)}
          />
        </label>
      </form>
      <Loader isLoading={isLoading} />
      {error && <ToastBox title={error && error.toString()} status="error" />}
    </>
  );
};

export default PatientsSearch;
