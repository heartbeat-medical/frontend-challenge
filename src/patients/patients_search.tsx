import { FunctionComponent, useEffect, useState } from "react";
import { Patient, PatientSearchQuery } from "./patients";
import { debounce } from "lodash";
import { useToast } from "../toast/ToastProvider";

type props = {
  loadPatients: (query: PatientSearchQuery) => Promise<Patient[]>;
  onResults: (ps: Patient[]) => void;
};

export const PatientsSearch: FunctionComponent<props> = ({
  loadPatients,
  onResults,
}) => {
  const [query, setQuery] = useState("");
  const [loading, setLoading] = useState(false);
  const { addToast } = useToast();

  useEffect(() => {
    makeRequest();
  }, [query]);

  const makeRequest = () => {
    setLoading(true);
    const sq: PatientSearchQuery = {
      name: query,
    };

    loadPatients(sq)
      .then((ps) => {
        onResults(ps);
        setLoading(false);
      })
      .catch((err) => {
        addToast({
          title: "Error searching patients",
          message: err,
          status: "error",
        });
        setLoading(false);
      });
  };

  const changeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
    setQuery(e.target.value);
  };

  const debouncedHandler = debounce(changeHandler, 300);

  return (
    <div>
      <input onChange={debouncedHandler} aria-label="search-input" />
      {loading && <div>Loading...</div>}
    </div>
  );
};

type psearchboxprops = {
  onQueryChange: (query: PatientSearchQuery) => void;
};
