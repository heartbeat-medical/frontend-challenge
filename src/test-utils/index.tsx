import React from "react";
import { render, RenderOptions } from "@testing-library/react";

interface CustomRenderOptions extends RenderOptions {
  initialHistory?: string[];
}

const customRender = (ui: any, options: CustomRenderOptions = {}) => {
  const { initialHistory, ...rest } = options;

  // Provide any context that the components may be expecting
  const Wrapper: React.FC = ({ children }) => <>{children}</>;
  return render(ui, { wrapper: Wrapper, ...rest });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
