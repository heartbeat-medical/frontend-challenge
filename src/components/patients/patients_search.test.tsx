import { screen, render } from '@testing-library/react';
import { PatientsSearch } from './patients_search';
import userEvent from '@testing-library/user-event'
import { act } from 'react-dom/test-utils';

describe('<PatientsSearch />', () => {
  it('renders correctly', () => {
    const loadPatientsStub = jest.fn().mockResolvedValue([]);;
    const onResultsStub = jest.fn().mockResolvedValue(null);

    const { container } = render(
      <PatientsSearch 
        loadPatients={loadPatientsStub}
        onResults={onResultsStub} 
      />
    )

    expect(container).toMatchSnapshot();
  });

  it('searches for patients', async () => {
    const mockLoadPatients = jest.fn().mockResolvedValue([]);
    const stubOnResult = jest.fn().mockResolvedValue(null);
    render(
      <PatientsSearch
        loadPatients={mockLoadPatients}
        onResults={stubOnResult}
      />
    );

    const input = await screen.findByRole('textbox') as HTMLInputElement;

    await act(async () => {
      userEvent.type(input, 'Andre{enter}');
    });

    expect(input).toHaveFocus();
    expect(input).toHaveValue('Andre');
    expect(mockLoadPatients).toHaveBeenCalled();
  });
});
