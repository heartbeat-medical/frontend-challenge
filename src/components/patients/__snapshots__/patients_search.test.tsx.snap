// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PatientsSearch /> renders correctly 1`] = `
<div>
  <div>
    <div
      class="loading-container"
    >
      <div
        class="loading-spinner"
      />
    </div>
    <input />
  </div>
</div>
`;
